src/snowflake/snowpark/_internal/temp_table_auto_cleaner.py

```diff
@@ -0,0 +1,95 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+import logging
+import weakref
+from collections import defaultdict
+from queue import Empty, Queue
+from threading import Event, Thread
+from typing import TYPE_CHECKING, Dict, Optional
+
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import SnowflakeTable
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # pragma: no cover
+
+
+class TempTableAutoCleaner:
+    """
+    Clean up temporary tables created automatically by Snowpark at the session level.
+
+    This function identifies and removes any temporary tables that have been
+    generated by Snowpark during the current session. Temporary tables are
+    typically used for intermediate computations (e.g., df.cache_result) and
+    are not needed when they are not referenced anymore. Removing these tables
+    helps maintain a clean working environment and reduce storage cost for a long-running
+    session.
+    """
+
+    def __init__(self, session: "Session") -> None:
+        self.session = session
+        # this dict maintains key-value pair from Snowpark-generated temp table fully-qualified name
+        # to its reference count for later temp table management
+        # this dict will still be maintained even if the cleaner is stopped (`stop()` is called)
+        self.ref_count_map: Dict[str, int] = defaultdict(int)
+        self.queue: Queue = Queue()
+        self.cleanup_thread: Optional[Thread] = None
+        self.stop_event = Event()
```

Comment:
comment for stop event? that seems a single state, what is the benefit of using event, not just a single boolean flag ?